implementation/src/main/java/io/smallrye/jwt/auth/principal/KeyLocationResolver.java

```diff
@@ -56,146 +58,124 @@
+        LOGGER.debugf("Trying the create a key from the local JWK(S)...");
+
+        try {
+            return getKeyFromJsonWebKeys(kid, localJwks);
+        } catch (Exception e) {
+            LOGGER.debug("Failed to create a key from the local JWK(S)", e);
         }
 
         return null;
     }
 
-    PublicKey getContentJwk(String content, String kid) throws JoseException {
-        JsonObject jwk = null;
-
-        JsonObject jwks;
-        try (JsonReader reader = Json.createReader(new StringReader(content))) {
-            jwks = reader.readObject();
-        }
-        JsonArray keys = jwks.getJsonArray(JsonWebKeySet.JWK_SET_MEMBER_NAME);
-
-        if (keys != null) {
-            if (kid != null) {
-                for (int i = 0; i < keys.size(); i++) {
-                    JsonObject currentJwk = keys.getJsonObject(i);
-                    if (kid.equals(currentJwk.getString(JsonWebKey.KEY_ID_PARAMETER, null))) {
-                        jwk = currentJwk;
-                        break;
-                    }
-                }
-            } else if (keys.size() == 1) {
-                jwk = keys.getJsonObject(0);
+    private static void verifyKid(JsonWebSignature jws, String expectedKid) throws UnresolvableKeyException {
+        if (expectedKid != null) {
+            String kid = getKid(jws);
+            if (kid != null && !kid.equals(expectedKid)) {
+                LOGGER.debugf("Invalid token 'kid' header: %s, expected: %s", kid, expectedKid);
+                throw new UnresolvableKeyException("Invalid token 'kid' header");
             }
-        } else if (kid == null || kid.equals(jwks.getString(JsonWebKey.KEY_ID_PARAMETER, null))) {
-            jwk = jwks;
-        }
-
-        if (jwk != null) {
-            return PublicJsonWebKey.Factory.newPublicJwk(jwk.toString()).getPublicKey();
         }
+    }
 
-        return null;
+    private static String getKid(JsonWebSignature jws) throws UnresolvableKeyException {
+        return jws.getHeaders().getStringHeaderValue(JsonWebKey.KEY_ID_PARAMETER);
     }
 
-    private void loadContents() throws IOException {
+    private void initializeKeyContent() throws Exception {
         final String keyLocation = authContextInfo.getPublicKeyLocation();
         if (keyLocation.startsWith(HTTPS_SCHEME)) {
+            LOGGER.debugf("Trying to load the keys from the HTTPS JWK(S)...");
             httpsJwks = new HttpsJwks(keyLocation);
             httpsJwks.setDefaultCacheDuration(authContextInfo.getJwksRefreshInterval().longValue() * 60L);
             return;
         }
 
+        String content = readKeyContent(keyLocation);
+
+        // Try to init the verification key from the local PEM or JWK(S) content
+        verificationKey = tryAsPEMPublicKey(content);
+        if (verificationKey == null) {
+            verificationKey = tryAsPEMCertificate(content);
+        }
+        if (verificationKey == null) {
+            localJwks = loadLocalJwks(content);
```

Comment:
This should guard against the content not being valid JWK / JWKS. The `loadLocalJwks` and `getLocalJwk` methods will probably throw in such case, and we should catch here and do some debug logging.

And we should perhaps check in this class constructor if `verificationKey` is still `null` after `initializeKeyContent` returns. Currently, the constructor expects `initializeKeyContent` to either set the `verificationKey` field, or throw. Some methods, however, don't throw, while some do. This is very hard to follow.

Reply:
Alternatively, we should let all methods just throw.

Reply:
Actually my comment above isn't entirely accurate; I'm aware of that. This entire class is very very hard to understand.

Reply:
I'm gonna say one thing: the common wisdom is that many small functions are better than one single big function. The common wisdom is a piece of crap. The functions are scattered around with no order, their names are mildly confusing, and the control flow is far from obvious.

Reply:
OK, I went through the class in an IDE and it looks mostly fine, but I'm not sure about `loadLocalJwks`. I think it should [attempt to] never throw, and instead return `null`. That's what all methods in this class do.

Reply:
@Ladicek, thanks for giving it a thorough check :-). Yes, I was conscious as well about various methods mixed in, so at least I tried to push all the static methods at the end of the class. My main target was to optimize the creation of the keys.  Pushed some code into the standalone functions like reading the input stream, to simplify the testing.
Going to make sure a `null` is returned, indeed it would be more consistent